### ISSUE TITLE
[Feature Request/bug] flipping 2D characters in godot 3.0

### ISSUE THREAD

**Issue by aaronmzn** [godotengine/godot#12335]
Hello, so I wanted to try out 3.0 and began porting a game prototype I had in godot 2.1.4 and got a ridiculous "bug?"
![somethingwrong2](https://user-images.githubusercontent.com/15176623/31867076-4551320a-b75f-11e7-9eac-8f8126759a9a.gif)

This was working fine in godot 2.1.4. Flipping the main node of the character (AKA KinematicBody2D). 

but why like that? so all the stuff that are child of the character will flip with the character too, like weapons scenes with its own hitbox and script, particle effects, more sprites, etc. In Godot 2.1.4 actually worked really well without any physics problem as long as you don't use weird numbers.

the way I did it was changing the node scale of the main node (KinematicBody2D) from (1, 1) facing right to (-1, 1) facing left.
 
Feature Request: this is not crazy needed but maybe a Flip H and Flip V checkbox in the physicsbodies like in the Sprite node. that way you don't have to flip every single child on the character and you don't mess with the scale numbers. at be able of doing it of code or fixing the bug I'm having

I know you guys will say you're not supposed to change the scale of a physicsbody but to me it makes
more sense to flip the main node of the character instead of the sprite node or any other node.

having all the weapons and stuff as a child of the sprite node doesn't feel right and at the same time even if thats the way of doing it, you have to tell new people that that's the way of doing it.

**Issue description:**
the characters goes crazy and flips H and V very fast when flipping the character (that didn't happen in 2.1.4)

**Steps to reproduce:**
when moving to the left flip the main node with set_scale((-1,1) in script then ingame move to the left and it happens

**Link to minimal example project:**
[platformTEST.zip](https://github.com/godotengine/godot/files/1405432/platformTEST.zip)
godot 3.0 build 2017-10-22

**Comment by tagcup** [comment]
A non-uniform scale like (-1,1) isn't supposed to work with godot's physics code (or bullet).
If it did work at some point, it's was due to an accident or bug.

**Comment by Zylann** [comment]
You could workaround this by flipping only the "visual" nodes of your character.

**Comment by DanielKinsman** [comment]
Yes except characters aren't always symmetrical. You might get away with it for a person, but not a centaur. Or a person holding a gun as in @aaronmzn's example

One way of doing this currently is saving four separate transforms for each child (normal, hflipped, vflipped, hflipped+vflipped) and picking between them in code on the fly. It's a bit messy and would fall over for non-symmetric children like a custom drawn ConvexPolygonShape2D though. Perhaps somebody knows a better way? An out of the box solution or a plugin would be great.

**Comment by Dillybob92** [comment]
I did change your set_facing code to

```
var old_facing
func set_facing( dir ):
	facing = dir
	if old_facing != facing:
		set_scale( Vector2( facing , 1 ) )
		old_facing = facing
```

and it fixed the twitchy behavior (not for vertical, not sure why your char goes down)

anyway, as @Zylann said, just put the kinematic body in a node2d (container) and flip that

**Comment by DanielKinsman** [comment]
> anyway, as @Zylann said, just put the kinematic body in a node2d (container) and flip that

That produces very odd results, at least in 2.1.4. Flipping (via scale) a Node2D containing a RigidBody2D will work for a fraction of a second before it kicks itself back to the original orientation.
[flipping_hell.tar.gz](https://github.com/godotengine/godot/files/1406714/flipping_hell.tar.gz)

Be sure to turn on visible collision shapes when running this.

**Comment by Dillybob92** [comment]
OHHH. @DanielKinsman. Oops, I was using KinematicBody! My bad sir, you are correct. RigidBody does have that effect.

**Comment by Zylann** [comment]
@Dillybob92 @DanielKinsman I never said to put KinematicBody under a Node2D^^" I proposed to flip visuals instead, so basically not flipping the root node, as the root node must very often be the physics body.

**Comment by rtsketo** [comment]
This bug is killing me 😋 
Did it exist in older builds too?
How long does it usually take to fix a reported bug?

**Comment by aaronmzn** [comment]
by the way if you delete the code to make it flip and you edit the scaling in the editor to (-1,1) or set it on _ready() function it works fine without going crazy but ofcourse stays like that. 
maybe its because is in physics_process. If someone can do it so it runs the script once when you press left or right (like with signals maybe) and see what happens, because I'm not skilled enough in coding

and for putting it as a child of a Node2D it works kind of but when you walk you keep getting way from the origin position of the node2d and when you go to the other way you teleport to the negative x of the current position like fliping a page of a book.

well I hope this gets fixed, this is a must for fighting games for example. even for top-down games too.

![zangief_sfa_various_hit_boxes_display](https://user-images.githubusercontent.com/15176623/31923068-1461af26-b84f-11e7-8285-a52a72715a3e.png)

**Comment by rtsketo** [comment]
Oh I see. So a quick workaround would be duplicating the KinematicBody, setting the scale to (-1,1) for one of them and then switching visibility depending on which way the characters looks at.

**Comment by Zylann** [comment]
@rtsketo If your collision shapes are not symmetrical, then no, there is no way to flip them with scaling. An alternative would be to toggle, rotate or offset them instead. Or, simplest: not make them assymetrical in the first place, there are alternatives. And again, I don't know what's wrong with Node2D but when I say "flipping visuals" I am referring to such node structure:

- KinematicBody
    - CollisionShape1 <-- do something else with these
    - CollisionShape2
    - visuals (Node2D) <-- flip only this
        - Sprite
        - Particles

**Comment by DanielKinsman** [comment]
I've just done exactly that, here it is for everyone who wants to use it: https://github.com/DanielKinsman/godot-tricks

**Comment by aaronmzn** [comment]
@DanielKinsman for some reason it doesn't work for me, it shows an error on the line 32,13 if not (n extends Node2D):
_expected ')' in expression_ on FlipHelper.gd

**Comment by DanielKinsman** [comment]
Issues with my code should be discussed in [my repo](https://github.com/DanielKinsman/godot-flippable-physics/issues), not this one. I am using godot 2.x so you may have to adapt it for 3.x. I will probably update it myself once 3.x is released.

**Comment by aaronmzn** [comment]
Setting the size and position values for every child to make it flip seems unpractical and not every game needs to flip only the visuals to make it look like it changed direction.

I waited for godot 3.0 beta and the bug is still remaining, but **guys**, I noticed something that may help fix this. 

I only know this happens only on KinematicBody2D: 

When you are not using move_and_collide() or move_and_slide() in the script this method _(setting the scale to (-1,1))_  works fine, but when you use either of them then it happens. But only when setting the size.X value to a negative value. Doing it with Y works fine and **I think I know what is causing this**.

When using move_and_collide() or move_and_slide() and you set the size.x to a negative value like -1, the bug happens, and **it sets the rotation_degree to -180 only when X is negative.**

This may seem like something that is not required to get fixed but if you stop and think about it, this is needed for big bosses where you need to put multiple collision shapes, or in a fighting game like I said before.

Please, if any godot developer can fix this in base of what I said it would be great. I suggest to put an option to fliph and flipv like in the sprite node, that would be handy.

godot 3.0 beta1
[fliptest.zip](https://github.com/godotengine/godot/files/1549649/fliptest.zip)

**Comment by rtSketo** [comment]
It isn't tagged as a bug, so I think it won't get fixed. Correct?

**Comment by aaronmzn** [comment]
I guess, even that it isn't considered a bug, then it can be considered a feature request. If I knew how to do it myself I would fixed it or add the feature to flip horizontally and vertically with a pull request but I don't have that knowledge but I do believe someone will need this when they reach the point I am.

The main request is to be able to flip the main node so every child node flips, but without affecting how the code works. Don't want to be too pushy but if the final godot 3.0 release still has this issue some people will encounter the same problem that I had when they try to do the same thing.

**For now I found the way of fixing it anyway**, set rotation_degree to 0 after setting scale.x to -1, this is in case someone got the same issue. Now I can start porting a project I'm working on 2.1.4 to 3.0 😄

**Comment by puppetmaster-** [comment]
To visual whats going on I created a test scene in godot2 and godot3.

## Godot2
![godot2_kinematicbody2d_transformation](https://user-images.githubusercontent.com/10463138/36076354-5aadfc10-0f5b-11e8-9c78-d446852efc66.gif)
- the global_transformation.get_scale() don't change and stays at 1,1
- when move(Vector2(0,0)) is used KinematicBody2D.get_scale() stays at 1,1 but visually it changes
- no -0 value in transformation

## Godot3
![godot3_kinematicbody2d_transformation](https://user-images.githubusercontent.com/10463138/36076609-d57cd684-0f5e-11e8-8356-16a5b881d07d.gif)
- -0 value in transformation
- global_transformation.get_rotation() does not remain 0
- move_and_slide(Vector2(0,0)) or move_and_collide(Vector2(0,0)) change global_transformation
- object.get_scale() and global_transformation.get_scale() are not the same

## QuickFix
to flip a KinematicBody2D in Godot3 you need to set the global_transformation inside the _physics_process()
- scale(1,1) = set_global_transform(Transform2D(Vector2(1,0),Vector2(0,1),Vector2(position.x,position.y)))
- scale(-1,1) = set_global_transform(Transform2D(Vector2(-1,0),Vector2(0,1),Vector2(position.x,position.y)))

![godot3_kinematicbody2d_flip](https://user-images.githubusercontent.com/10463138/36076804-7d947ff0-0f61-11e8-9749-a846a8035a10.gif)

I'm not the right person to analyze the behavior in godot3 but maybe I was able to provide some valuable information about it.

**Comment by Arkhano** [comment]
I have a similar bug but without a RigidBody2D, I used my own physics. 

So if you animate a collisionShape2D, and you try to flip the parent Node2D with scale.x, the Animation Player will make the characters goes crazy and flips H very fast while you move to the left. The only fix for this for me at this moment is to make all the animation to the left too. =/

Please please, fix this or at least make a flip_h or flip_y for the CollisionShape2D, cause scale it with -1 doesn't work the same way.

Remember fighting games or animations in general that needs to change the collider position needs this.

Thank you.

**Comment by eon-s** [comment]
@Arkhano bodies do not accept scaling because the state transform and node transform are somehow linked and the physics engine cannot handle scales.
But you can scale the visual representation (sprite) and collision shape, it may be enough in most cases.

It will need a more elegant way to manage scales, though, because more complex things like mesh based figures may need a lot of corrections or special body setup to flip objects.

**Comment by Arkhano** [comment]
@eon-s The problem is if you do an animation to the right where the position of the sprite and collisionShape2D are for example x, x + 1 and x + 2 and you scale the collision and sprite to -1, Animation Player will continue doing the animation to the right and not doing the corresponding animation to the left, it flips the sprite but only reverse (not flip) the vertices of collisionShape2D. 

The only solution I know for this right now is to make all the animations for the left side too.

**Comment by arthurpaulino** [comment]
For me, it worked with `scale.x = scale.y * direction`, where `direction` equals 1 if your character should face right and -1 otherwise.

**Comment by AttackButton** [comment]
> For me, it worked with `scale.x = scale.y * direction`, where `direction` equals 1 if your character should face right and -1 otherwise.
** scanle.x * direction, right?

Yeah, I make this way too in my game. The problem with this is it strangely **flip the coordinate sign also of it children in the process**, and in specifics situations (like when you make if statements that compares the local (fliped) positions with the global ones),  you end up needing to create a whole new logic just to deal with it. I learned how to workaround to use it, **but it stills strange to have positives X sign coordinates that goes to the left and negatives ones going to the right when you flip it.** 

I think the problem of https://github.com/godotengine/godot/issues/4263 is somehow relative with this one.

**Comment by arthurpaulino** [comment]
> ** scale.x * direction, right?

Nope, it's `scale.x = scale.y * direction`. I'm using Godot 3.1.2.

**Comment by giovanitrentin** [comment]
> > ** scale.x * direction, right?
> 
> Nope, it's `scale.x = scale.y * direction`. I'm using Godot 3.1.2.

It´s Work! Great!

**Comment by luislodosm** [comment]
If the scale of the KinematicBody2D node is not (1,1):
```
	if velocity.x > 0:
		scale.x = scale.x * sign(scale.y)
	elif velocity.x < 0:
		scale.x = -scale.x * sign(scale.y)
```
It's weird. (Godot 3.2.1)

**Comment by KaizNike** [comment]
I couldn't find an amiable solution to this, so I just set all directional collision to be children of my sprite node, which flips normally in the code. This feels like a cheap solution.

**Comment by seattleowl** [comment]
If anyone is still here, you can use flip_v and flip_h on the sprite.

**Comment by kronok** [comment]
> If anyone is still here, you can use flip_v and flip_h on the sprite.

It's not about just the visuals, though, it's about the collisions, etc. too.

**Comment by rayxuln** [comment]
I use this and it work fine for me:

```
func flip_h(flip:bool):
	var x_axis = global_transform.x
	global_transform.x.x = (-1 if flip else 1) * abs(x_axis.x)
```

**Comment by edsulaiman** [comment]
After doing a simple test I concluded that.
- when we change the value of scale x to -1, the value of scale x will become -1.
- but if the node position changed, the scale x  value will be reset to 1 again.

code:
``` python 
func _physics_process(delta):
	if (Input.is_action_just_pressed("ui_left")):
		scale.x = -1
	if (Input.is_action_just_pressed("ui_right")):
		scale.x = 1

	print(scale.x)

	var xValue = Input.get_action_strength("ui_right") - Input.get_action_strength("ui_left")
	move_and_slide(Vector2(xValue, 0))
```
print result when node position is not changed:
```
1
1
-1 <--- When ui_left just pressed
-1
-1
```

print result when node position changed:
```
1
1
-1 <--- When ui_left just pressed
1 <--- Reset to 1 after node position changed
1
```

## Additional information
- Godot Engine v3.3.2 mono

**Comment by DogTorrent** [comment]
I am using RigidBody2D(mode is Character) and I have tried all the solution of above but none of them can solve the problem now...
the only way to flip it horizontally is to change scale.x of its children:
```
for child in get_children():
	if child.get("scale") != null:
		child.scale.x = -child.scale.x
```
NOT ELEGANT

I find #5734 but i think we can at least make a flip_h/flip_v function for RigidBody2D(or anything else with the same problem)

Version: 3.3.2 stable

**Comment by berarma** [comment]
If this isn't considered a bug but normal behavior, why isn't it documented anywhere?

The documentation leads you to believe you can animate any property on any node, and the editor let's you do it without any warning. I think this should be considered a bug in the documentation, editor or engine.

I'd like to be able to help fixing it but I'm new to Godot. These unexpected traps are the only negative in Godot.

**Comment by Calinou** [comment]
I'm not sure where this should be documented exactly. Should we add a note to both RigidBody2D and KinematicBody2D? Does this also affect StaticBody2D?

**Comment by pouleyKetchoupp** [comment]
> I'm not sure where this should be documented exactly. Should we add a note to both RigidBody2D and KinematicBody2D? Does this also affect StaticBody2D?

Yeah, possibly a note in the documentation for RigidBody2D and KinematicBody2D until it is supported, although I'm not sure it would be easy to find. I don't think it affects StaticBody2D in any way.

Otherwise we could add an editor warning when the scale is set on KinematicBody2D (the same way it's already done for RigidBody2D), and maybe some runtime errors on local transform changed as well.

**Comment by eon-s** [comment]
Some things are mentioned in the Physics tutorials
https://docs.godotengine.org/en/stable/tutorials/physics/physics_introduction.html#collision-shapes
https://docs.godotengine.org/en/stable/tutorials/physics/kinematic_character_2d.html#scene-setup

And the RigidBody2D reference
https://docs.godotengine.org/en/stable/classes/class_rigidbody2d.html#description

But maybe it need a separate entry in the physics tutorials that talk about all these "peculiarities" of the engine (shape/body scaling, collision reporting of mask-layers pairs on 3.x, limit with planes, segment shapes, etc.).

--------------

The OP issue may need better UX approach, maybe automagic resizes of shapes when trying to scale bodies, I don't know if something like this is planned for 4.

**Comment by DinkleDorph** [comment]
7 year old thread and this is still an issue 😬 . The solution posted by arthur does work though.

**Comment by zentaki** [comment]
I get completely the OP.
When we design a character for a 2D we think of one side and just flip it; The only problem here is that with non symmetrical shapes will cause clipping. So character and it's features should be done in such design that maintain symmetry so the visual and collision work together.
Lets say I have a character with non symmetrical shape, then if I am already colliding with the wall with the minor side then I flip the character, the bigger side will clip through the wall, probably making it stuck.
But that should not be a reason not to feature godot to flip the entire object.
yeah that clipping can occur with whatever you might add into the character for the same reason the other clipping does.

I guess this is why they did not put such a feature in godot, and must have other strategies which are not so obvious to deal with such things.
Maybe would be easier to have the feature and deal with the clipping later. 

In the 3D we do have same problem, not for turning character around and clipping, but for clipping the weapons on walls. And to solve that is also not so simple at first. A character that has a gun, will shoot bullets at the tip of the gun. IF you collide the gun to the wall you solve the clipping problem but make the game collisions against wall looks weird, but if you leave as it is you would be able to shoot through walls which is mostly game breaking and undesirable, One way you break the immersion feels the other you break the mechanics. So it must be solved in a clever way, If I recall is solved with some kind of illusion that does not break the immersion feel but makes sure that the mechanic does not break either.

Anyway, it seems that we should also do some more work to make it happen nice with a 2D world. So flip the image as you would, but when you do that you also have flip the other stuffs, guess it can be made using position changings, or using 2 objetos that represent 2 points and you change the point to be used depending on the side you are facing. I think overall trying to use scale flip will mostly break things apart. 
For animation and collisions you can use animation player to do the work if I am not mistaken. I am sure is as simple as the animation itself to set up a changing collision according to the animation, and as a thumb of rule, is always good to have symmetrical shapes, otherwise things will get really hard to do.

But yes it does sucks not have a simple way like just flipping the whole object instead. I could make things simpler but in the other hand you would have to deal with the consequences later anyway

**Comment by mateuszfryc** [comment]
Version 4.4.1 here and I can confirm that even though flipping CharacterBody2D's scale still misbehaves the solution with global_transform.x.x works:

```
extends CharacterBody2D

var faces_right := true

func _physics_process(delta):
	var input_vector = Vector2(
		Input.get_action_strength("Move_Right") - Input.get_action_strength("Move_Left"),
		Input.get_action_strength("Move_Down") - Input.get_action_strength("Move_Up")
	).normalized()

	if input_vector.x > 0.7 and !faces_right:
		faces_right = true
		if global_transform.x.x != 1.0:
			global_transform.x.x = 1.0
	elif input_vector.x < -0.7 and faces_right:
		faces_right = false
		if global_transform.x.x != -1.0:
			global_transform.x.x = -1.0
```

**Comment by eon-s** [comment]
This is still a UX issue, by the way Godot works, bodies shouldn't be scaled, the workaround is to have a flip function that:
- Executes flipX in the sprites/visuals
- "mirrors" the offsets of colission shapes for offsetted shapes

The solution may involve a "flip" property for bodies, which flips selected children in a list.

**Comment by mateuszfryc** [comment]
The reason why some of us searched for a way to flip the entire character is that it is not just visuals that we needed to flip. My character for example has few other sub character nodes that are positioned in relative manner and if I would flip only the sprite I would need to manually flip all other nodes that should be always in front or in the back of the character. Example - hit boxes for character weapons in my game are always in front of the character.

In other type of games and even engines there are day to day cases when we rotate our character with all of its components. In 2D we often flip instead of rotating, so having a way to flip the entire character with its sub nodes makes perfect sense.
